Userland/Libraries/LibGfx/TGALoader.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2022, Tom Needham <06needhamt@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/String.h>
+#include <LibGfx/TGALoader.h>
+
+namespace Gfx {
+
+enum TGADataType : u8 {
+    None = 0,
+    UncompressedColorMapped = 1,
+    UncompressedRGB = 2,
+    UncompressedBlackAndWhite = 3,
+    RunLengthEncodedColorMapped = 9,
+    RunLengthEncodedRGB = 10,
+    CompressedBlackAndWhite = 11,
+    CompressedColorMapped = 32,
+    CompressedColorMappedFourPass = 33
+};
+
+struct TGAHeader {
+    u8 id_length;
+    u8 colour_map_type;
+    enum TGADataType data_type_code;
+    i16 colour_map_origin;
+    i16 colour_map_length;
+    u8 colour_map_depth;
+    i16 x_origin;
+    i16 y_origin;
+    u16 width;
+    u16 height;
+    u8 bits_per_pixel;
+    u8 image_descriptor;
+} __attribute__((packed));
+
+static_assert(sizeof(TGAHeader) == 18);
+
+union TGAPixel {
+    struct TGAColor {
+        u8 blue;
+        u8 green;
+        u8 red;
+        u8 alpha;
+    } components;
+
+    u32 data;
+} __attribute__((packed));
+
+static_assert(sizeof(TGAPixel) == 4);
+
+class TGAReader {
+public:
+    TGAReader(u8 const* data, size_t data_length)
+        : m_data(move(data))
+        , m_data_remaining(move(data_length))
+    {
+    }
+
+    ALWAYS_INLINE u8 read_u8()
+    {
+        VERIFY(m_data_remaining >= 1);
+        m_data_remaining--;
+        return *(m_data++);
+    }
+
+    ALWAYS_INLINE i8 read_i8()
+    {
+        return static_cast<i8>(read_u8());
+    }
+
+    ALWAYS_INLINE u16 read_u16()
+    {
+        return read_u8() | read_u8() << 8;
+    }
+
+    ALWAYS_INLINE i16 read_i16()
+    {
+        return read_i8() | read_i8() << 8;
+    }
+
+    ALWAYS_INLINE u32 read_u32()
+    {
+        return read_u16() | read_u16() << 16;
+    }
+
+    ALWAYS_INLINE i32 read_i32()
+    {
+        return read_i16() | read_i16() << 16;
+    }
+
+    ALWAYS_INLINE TGAPixel read_pixel(u8 bpp)
+    {
+        auto pixel = TGAPixel();
+
+        switch (bpp) {
+        case 24:
+            pixel.components.blue = read_u8();
+            pixel.components.green = read_u8();
+            pixel.components.red = read_u8();
+            pixel.components.alpha = 0xFF;
+            return pixel;
+
+        case 32:
+            pixel.components.blue = read_u8();
+            pixel.components.green = read_u8();
+            pixel.components.red = read_u8();
+            pixel.components.alpha = read_u8();
+            return pixel;
+
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    size_t data_remaining() const
+    {
+        return m_data_remaining;
+    }
+
+private:
+    u8 const* m_data { nullptr };
+    size_t m_data_remaining { 0 };
```

Comment:
Using a raw pointer for an array is a bit spooky. Could we use a `Span<u8>` for the data to make sure we never read of the edge of it, and then flip data-remaining to instead be an index? I guess you'd lose some speed there but we generally prioritise safety and readability over performance, at least until it's measured.

Reply:
Sounds like a good idea to me. I'm still getting used to the data structures in the project. will update

Reply:
Done!